### PR TITLE
Changed authentication method

### PR DESCRIPTION
**Pull request by lucaleivaloop**
#### Description:

Change the gem used to produce the auth token.

---

@loopstudio/ruby-devs

### CODE REVIEW

File: app/models/auth_token.rb [hunk 1]
```diff
@@ -1,17 +1,14 @@
 class AuthToken
-  def self.key
-    ENV.fetch('RAILS_MASTER_KEY')
-  end
+  KEY = OpenSSL::PKey::RSA.generate 2048
```
Comment (letiesperon) [inline app/models/auth_token.rb:2]: use parenthesis for method calls

File: app/models/auth_token.rb [hunk 2]
```diff
@@ -1,17 +1,14 @@
...
   def self.token(user)
     payload = { user_id: user.id }
-    JsonWebToken.sign(payload, key: key)
+    JWT.encode(payload, KEY, 'RS256')
   end
 
   def self.verify(token)
-    result = JsonWebToken.verify(token, key: key)
-    return nil if result[:error]
+    result = JWT.decode(token, KEY.public_key, true, { algorithm: 'RS256' })
```
Comment (letiesperon) [inline app/models/auth_token.rb:10]: avoid calling variables `result`, it is not nemotecnic. technically, all method calls are a result

File: app/models/auth_token.rb [hunk 3]
```diff
@@ -1,17 +1,14 @@
...
 
-    User.find_by(id: result.dig(:ok, :user_id))
+    User.find_by(id: result.first['user_id'])
```
Comment (letiesperon) [inline app/models/auth_token.rb:12]: why not dig? what happens if the decodification fails? does it raise some error?

File: app/models/auth_token.rb (hunk 1 above)
Comment (mconiglio) [reply app/models/auth_token.rb:2]: won't this change every time the server is restarted? Maybe we should keep this as an env variable

File: app/models/auth_token.rb (hunk 3 above)
Comment (lucaleivaloop) [reply app/models/auth_token.rb:12]: It does raise an error. How would you use dig on an array? `result.dig(:first, 'user_id')`?

File: app/models/auth_token.rb (hunk 2 above)
Comment (lucaleivaloop) [reply app/models/auth_token.rb:10]: I did not rename it, but I will, do you have a name in your mind?

Comment (diebas) [reply app/models/auth_token.rb:10]: 'decoded_token' maybe?

File: app/models/auth_token.rb (hunk 3 above)
Comment (letiesperon) [reply app/models/auth_token.rb:12]: yeah, you're right about dig

File: app/models/auth_token.rb (hunk 2 above)
Comment (lucaleivaloop) [reply app/models/auth_token.rb:10]: It has some extra info, but that might do I guess

File: .env.test [hunk 4]
```diff
@@ -0,0 +1 @@
+TOKEN_KEY=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```
Comment (mconiglio) [inline .env.test:1]: can't this be configured in the github secrets? or we need it anyway for testing locally?

File: .env.template [hunk 5]
```diff
@@ -1,5 +1,6 @@
 MAX_DEPTH=5
 MAX_COMPLEXITY=30
+TOKEN_KEY=
```
Comment (mconiglio) [inline .env.template:3]: maybe JWT_ENCODING_KEY? or something like that?

File: .env.test (hunk 4 above)
Comment (lucaleivaloop) [reply .env.test:1]: We need it for testing locally, we shouldn't use the same on prod staging nor dev servers though.

File: .env.test [hunk 6]
```diff
@@ -0,0 +1 @@
+JWT_ENCODING_KEY=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```
Comment (letiesperon) [inline .env.test:1]: why is this needed? cant we just do a dummy test key?

Comment (lucaleivaloop) [reply .env.test:1]: It has to be a valid `HS256` key.

Comment (mconiglio) [reply .env.test:1]: are you sure? in the gem readme example it's using `'my$ecretK3y'`

File: README.md [hunk 7]
```diff
@@ -49,6 +49,9 @@ For detecting N+1 issues.
 #### graphql-batch
 For allowing queries to be batched to avoid the N+1 issues. You can read the documentation [Here](https://github.com/Shopify/graphql-batch#usage).
 
+#### JWT
+To generate the tokens required to authenticate we use [JWT](https://github.com/jwt/ruby-jwt) it is required to set the ENV variable `JWT_ENCODING_KEY` with the result of running `OpenSSL::Digest.new('sha256')` on the rails console.
```
Comment (mconiglio) [inline README.md:53]: change sha256 to hmac

File: .env.test (hunk 6 above)
Comment (lucaleivaloop) [reply .env.test:1]: My mistake, it does work with a dummy, changing it.

File: README.md (hunk 7 above)
Comment (lucaleivaloop) [reply README.md:53]: If I run that it throws an error saying `hmac` is not supported.

Comment (letiesperon) [inline README.md:53]: I would say

```
We use [JWT]() for authentications.
```

And then I would add an extra step in the Setup list


WDYT?
